### PR TITLE
invokables: transparently use a CppObj wrapper

### LINKED ISSUE #8: Transparently use the CppObjWrapper in invokables and other methods
Currently we have `cpp: Pin<&mut CppObj>` as the type to signify the Cpp instance. This works until you try to perform multiple mutations, and doesn't allow us to perform the Rust <-> C++ conversions transparently. This also means users can perform tasks that might be considered "unsafe" - eg triggering an invokable via the C++ side rather than Rust (could cause deadlock).

Currently users need to write `let wrapper = CppObjWrapper::new(cpp);` and then user this wrapper.

Instead use something like `cpp: CppObj` or `cpp: CppObjWrapper` as the type the user uses, then transparently generate another rust method which is called first and takes the `Pin<&mut CppObj>`, then creates the wrapper and calls the users method.

### PR DESCRIPTION
**Pull request by ahayzen-kdab**
- [X] Add CppObj type
- [x] Use typedef on C++ so we have CppObj there as well (simplifies generation and allows custom Cpp class naming)
- [x] Support external modules for CppObj in inovkables
- [x] Support external modules for CppObj in properties ? (currently this is Ptr)
- [x] Support CppObj for traits (HandlePropertyChange and HandleUpdateRequest)
- [x] Remove unused types and code (Ptr and Pin ?)

Closes #8

### CODE REVIEW

File: examples/basic_cxx_qt_qml/src/data.rs [hunk 1]
```diff
@@ -27,9 +27,8 @@ mod my_data {
 
     impl RustObj {
         #[invokable]
-        fn as_json_str(&self, cpp: Pin<&mut CppObj>) -> String {
-            let wrapper = CppObjWrapper::new(cpp);
-            let data = Data::from(&wrapper);
+        fn as_json_str(&self, cpp: &mut CppObj) -> String {
+            let data = Data::from(&*cpp);
```
Comment (ahayzen-kdab) [inline examples/basic_cxx_qt_qml/src/data.rs:31]: At the moment we only implement `impl<'a> From<&CppObj<'a>> for Data`, so you can't do `Data::from(cpp)` here as that is `&mut CppObj`.

This creates not nice syntax of having to do `&*` to create a non-mut ref (?), so should we implement both `&CppObj` and `&mut CppObj` in the `From` ?

**Comment by ahayzen-kdab** [comment]
The `CppObj` as a typedef on the C++ doesn't fully work

- QML complains about the type as it doesn't know what it is
- In the `unsafe extern "C++"` so far it appears that the `type NAME` needs to match the C++ name (?), could we generate the `CrateModuleFFICppObj` as a typedef on C++ side and would that be enough or is it before this stage?

**Comment by ahayzen-kdab** [comment]
A typedef for `CrateModuleFFICppObj` doesn't work (?), so looks like all QObject names need to be unique. Which is also true for QML anyway, so maybe this is fine ? The problem becomes if we allow renaming of C++ classes.

**Comment by ahayzen-kdab** [comment]
QML with the typedef now works, so sub objects are now working in invokables without Pin visible - but related to #19 collisions could occur.

**Comment by ahayzen-kdab** [comment]
Wonder if we'll ever want the Pin type in the future to allow the user to pass custom C++ types across ? :thinking: But they would likely need to do other interactions with cxx anyway and may go the route similar to our QVariant/QString etc anyway :thinking: 

Probably worth a bit of thought incase that is worth keeping.

**Comment by ahayzen-kdab** [comment]
This functionally is working, there just needs to be thought into how to handle the external module object names still. As we cannot use CppObj everywhere in the cxx bridge.

**Comment by ahayzen-kdab** [comment]
Also more code can be removed and simplified :-)

**Review by LeonMatthesKDAB — COMMENTED** [review]
I'm not sure how much I can comment on the actual generation code, as I'm not familiar with it, but the resulting API is nice, the duality of CppObj and RustObj should be easy enough to understand.

There are still some test failures on my machine:
```
The following tests FAILED:
	 11 - cxx_qt_gen_cargo_clippy (Failed)
	 45 - example_basic_cxx_qt_qml_myobject_test_valgrind (Failed)
	 47 - example_basic_cxx_qt_qml_mydata_test_valgrind (Failed)
	 49 - example_basic_cxx_qt_qml_qttypes_test_valgrind (Failed)
	 51 - example_basic_cxx_qt_qml_plugin_test_valgrind (Failed)
	 53 - example_minimal_cxx_qt_qml_myobject_test_valgrind (Failed)
	 55 - example_qml_with_threaded_logic_website_test_valgrind (Failed)
Errors while running CTest
```

File: examples/basic_cxx_qt_qml/src/data.rs (hunk 1 above)
Comment (LeonMatthesKDAB) [reply examples/basic_cxx_qt_qml/src/data.rs:31]: Hm, maybe we will have to implement the From trait twice :thinking: 
It feels like their should be an easier solution for this, but I can't think of one atm

File: cxx-qt-gen/src/extract.rs [hunk 2]
```diff
@@ -193,19 +200,30 @@ fn extract_qt_type(
...
-        namespace.push(ident_str.clone());
-
-        Ok(QtTypes::Ptr {
-            ident_str,
-            // TODO: do we need to track is_ref here?
-            ident_namespace_str: namespace.join("::"),
+    // If the first ident is crate, the last is CppObj, and we have more than two parts
+    // then we are an external CppObj
+    } else if idents.len() > 2
+        && idents.first().unwrap().to_string().as_str() == "crate"
```
[the remark this answers is not in the document]
Comment (ahayzen-kdab) [reply cxx-qt-gen/src/extract.rs:206]: Yup good idea, i think we should check it ends in CppObj. We need the following to be true

  * Starts with crate/super (if more than one part)
  * Ends with CppObj
  * Has one part or three+ parts

[the remark this answers is not in the document]
Comment (ahayzen-kdab) [reply cxx-qt-gen/src/extract.rs:208]: I've reported an issue to track adding `super::module::CppObj` support, as this branch is already complex and that needs more changes / is lower priority #44 .

File: examples/basic_cxx_qt_qml/src/data.rs (hunk 1 above)
Comment (ahayzen-kdab) [reply examples/basic_cxx_qt_qml/src/data.rs]: I've implemented the mutable From trait by calling the immutable From trait.

File: cxx-qt-gen/src/extract.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (LeonMatthesKDAB) [reply cxx-qt-gen/src/extract.rs:208]: Ah, okay, then the problem is a bit more complex than I expected.
I'm not sure, we even need "starts with crate/super" to be true though...
What if I want to use a QObject provided by another crate that also uses cxx-qt?
Best to add that list to #44 too though.

[the remark this answers is not in the document]
Comment (LeonMatthesKDAB) [reply cxx-qt-gen/src/extract.rs:208]: I've updated the issue :)

**Comment by ahayzen-kdab** [comment]
> I'm not sure how much I can comment on the actual generation code, as I'm not familiar with it, but the resulting API is nice, the duality of CppObj and RustObj should be easy enough to understand.
> 
> There are still some test failures on my machine:
> 
> ```
> The following tests FAILED:
> 	 11 - cxx_qt_gen_cargo_clippy (Failed)
> 	 45 - example_basic_cxx_qt_qml_myobject_test_valgrind (Failed)
> 	 47 - example_basic_cxx_qt_qml_mydata_test_valgrind (Failed)
> 	 49 - example_basic_cxx_qt_qml_qttypes_test_valgrind (Failed)
> 	 51 - example_basic_cxx_qt_qml_plugin_test_valgrind (Failed)
> 	 53 - example_minimal_cxx_qt_qml_myobject_test_valgrind (Failed)
> 	 55 - example_qml_with_threaded_logic_website_test_valgrind (Failed)
> Errors while running CTest
> ```

As discussed the clippy issues come from a newer version of clippy (fixes are here #48 ), valgrind issues are potentially similar and need investigation. For me with valgrind 3.15.0 from Ubuntu 20.04 it works.

**Review by LeonMatthesKDAB — APPROVED** [review]
As the clippy issue has been moved into #48 I think this can be merged now

**Pull request merged**